L81-core-cpp-jwt-token-audience.md

```diff
@@ -0,0 +1,128 @@
+L81: Custom Audience in JWT Access Credentials and Google Default Credentials
+----
+* Author(s): [Jiangtao Li](https://github.com/jiangtaoli2016), [Yihua Zhang](https://github.com/yihuazhang)
+* Approver: markdroth
+* Status: In Review {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: Core, C++
+* Last updated: 2021-05-26
+* Discussion at: https://groups.google.com/g/grpc-io/c/o0Kkwc2JDa4
+
+## Abstract
+
+In current gRPC C++ implementation, the audience field of JWT access token is
+always set to the service URL. It works for most of the cases. However, there
+are use cases where a user wants to override the default audience (service URL)
+and uses its own. This gRFC provides a way for an applications to override
+the default audience field.
+
+## Background
+
+Consider an application wants to use JWT Access Credentials or Google Default
+Credentials to connect to a regional endpoint, e.g., 
+`us-dialogflow.googleapis.com`, rather than the global endpoint 
+`dialogflow.googleapis.com`, the cloud service may reject the JWT because the
+auidence field was set to `us-dialogflow.googleapis.com`. In such scenarios,
+the application wants the ability to override the default audience field and
+set the auidence field of its choice.
+
+### Related Proposals: 
+
+NA
+
+## Proposal
+
+We propose to change the following C++ API. If the user_provided_audience
+field is empty, then service URL will be used as the auidence field, otherwise,
+the user_provided_audience will be used as the audience in the JWT.
+
+When the user_provided_audience field is set in Google Default Credentials,
+this field is only used when a service account JWT access credential is
+created. This field will be no-op if for example a compute engine credential
+is created.
+
+```
+/// Builds Service Account JWT Access credentials.
+/// json_key is the JSON key string containing the client's private key.
+/// token_lifetime_seconds is the lifetime in seconds of each Json Web Token
+/// (JWT) created with this credentials. It should not exceed
+/// \a kMaxAuthTokenLifetimeSecs or will be cropped to this value.
+/// user_provided_audience is an optional field for user to override the
+/// auidence in the JWT token. If user_provided_audience is empty, the service
+/// URL will be used as the audience.
+std::shared_ptr<CallCredentials> ServiceAccountJWTAccessCredentials(
+    const grpc::string& json_key,
+    long token_lifetime_seconds = kMaxAuthTokenLifetimeSecs,
+    const grpc::string& user_provided_audience = "");
+
+
+/// Builds credentials with reasonable defaults.
+/// user_provided_audience is an optional field for user to override the
+/// auidence in the JWT token. If user_provided_audience is empty, the service
+/// URL will be used as the audience.
```

Comment:
nit: we should probably note in this API comment that `user_provided_audience` will only be used if a service account JWT access credential is created by the application default credentials mechanism